spi: spi_ll_stm32: Fix spi_slave case in transceive()

avisconti opened the pull request:
If we are in SPI slave context we need to check, before
exiting the transceive() routine, whether the ret value
really represents a I/O error or instead is the number
of frames received.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
I realized that the git commit message was not clear at all. I re-formulated it and re-pushed the commit.

tbursztyka commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -414,6 +414,16 @@ static int transceive(struct device *dev,
+		return ret; /* number of frames received */
+	}
+#endif
+
```
Actually, I would have rewritten how ret is handled here in transceive: removing the below test for logging error, and removing the ret ? -EIO : 0; to me, in case of error, ret should already have the right error code.

tbursztyka replied:
you'll need to make the function returning errors giving directly the right error code with relevant changes depending on whether you are slave or master.

superna9999 replied:
I don't understand why this should be Slave specific, from API point of view Slave or Master transfers should be the same, with the same error return scheme.
ret != 0 means an error, ret = 0 mean you no error
why do you want the number of frames when an error occurs ?

tbursztyka replied:
This is part of the SPI changes that happened recently. In slave mode: if an error occurs it will need to return a negative error code, but in case of success it will have to return the number of received frames.
Master mode is as before: 0 on success, an negative error code otherwise.

avisconti replied:
As I commented few minutes ago after re-pushing the commit, the situation is now clean and aligned with other SPI controllers (like dw or intel), except when CONFIG_SPI_STM32_INTERRUPT is disabled. In that case I had to add a check that in all other cases is already done in the context part. 
Question: do we really need the stm32 case with interrupt disabled?

superna9999 replied:
My bad, I missed the point, sorry

avisconti committed:
spi: spi_ll_stm32: Fix transceive() ret value in spi_slave case

In SPI slave case the transceive should return either the negative
errno code in case of error or the number of frames received.
So, now:

 1. the spi_stm32_get_err() routine already checks whether
    the SPI cell got an error and returns -EIO in that case.

 2. the transceive() routine always returns whatever the
    spi_context_wait_for_completion() has returned, which
    is either:
        a. -EIO in case of error
        b. 0 in spi_master ok case
        c. the number of frames received in spi_slave ok case

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@tbursztyka 
I aligned the transceive() more or less to what the other cases (dw, intel, ...) are doing.
This when STM32 interrupts are enabled. If not, I had to add the test that usually is done inside the wait_for_completion context routine.

superna9999 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -52,7 +52,13 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
+	if (sr & SPI_STM32_ERR_MSK) {
+		SYS_LOG_ERR("%s: err=%d", __func__,
+			    sr & SPI_STM32_ERR_MSK);
+		return(-EIO);
```
Add a space before (-EIO)

superna9999 reviewed (approved):
With the small (-EIO), LGTM
